collector/pg_stat_progress_vacuum.go

```diff
@@ -0,0 +1,202 @@
+// Copyright 2025 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package collector
+
+import (
+	"context"
+	"database/sql"
+	"log/slog"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+const progressVacuumSubsystem = "stat_progress_vacuum"
+
+func init() {
+	registerCollector(progressVacuumSubsystem, defaultDisabled, NewPGStatProgressVacuumCollector)
+}
+
+type PGStatProgressVacuumCollector struct {
+	log *slog.Logger
+}
+
+func NewPGStatProgressVacuumCollector(config collectorConfig) (Collector, error) {
+	return &PGStatProgressVacuumCollector{log: config.logger}, nil
+}
+
+var (
+	statProgressVacuumPhase = prometheus.NewDesc(
+		prometheus.BuildFQName(namespace, progressVacuumSubsystem, "phase"),
+		"Current vacuum phase (numeric). Use label mapping externally for human-readable values.",
+		[]string{"datname", "relname"},
```

Comment:
We could make the phase a human readable string by using a "State Set" style mapping.

```suggestion
		prometheus.BuildFQName(namespace, progressVacuumSubsystem, "phase"),
		"Current vacuum phase (numeric). Use label mapping externally for human-readable values.",
		[]string{"datname", "relname", "vacuum_phase"},
```

Then we would have metrics like this:
```
pg_stat_progress_vacuum_phase{datname="postgres",relname="foo",vacuum_phase="initializing"} 0
pg_stat_progress_vacuum_phase{datname="postgres",relname="foo",vacuum_phase="waiting for checkpoint to finish"} 0
pg_stat_progress_vacuum_phase{datname="postgres",relname="foo",vacuum_phase="estimating backup size"} 1
...
```

This allows for easy boolean checks in PromQL.

Reply:
One consideration here was the size overhead of strings versus a numeric. I think the current approach should work well for our needs, but I'm happy to make this change if you feel more strongly about it and it aligns with other collectors 👍 .

Reply:
The size overhead for the strings is no big deal, Prometheus does internal string deduplication in the index these days.

From a query ergonomics perspective, the boolean state metrics are better.

Reply:
I've switched this to the "State Set" style mapping. Looks much better!